### ISSUE TITLE
Add Remote Services Admin (RSA)  management agent

### ISSUE THREAD

**Issue by scottslewis** [tcalmant/ipopo#60]
Part of the OSGi Enterprise specification is Remote Services Admin (chapter 122).  RSA is a management agent of OSGi remote services.  The management agent gives more fine-grained control of remote services, and allows services to be exported and imported by custom topology managers.  

RSA also supports a multi-distribution provider approach, allowing transport-independent remote services.

An description of some of the advantages of such a multi-distribution provider approach are described here
http://eclipseecf.blogspot.com/2016/04/transport-independent-remote-services.html

and ECF provides a spec-compliant java RSA implementation supporting multiple distribution providers

It would be very useful for the RSA functionality to be made available via iPopo so that iPopo remote service users would have:

1) The flexibility and extensibility defined by the RSA specification
2) Have the ability to support multiple distribution providers, for multi-language remote services and other reasons.

I would like to contribute an RSA implementation to iPopo.  I'm the implementer of the ECF RSA implementation and this will make it easier to do so in python/iPopo.  It may require some refactoring of the existing iPopo remote services, but I think functional backward compatibility will/can be maintained.

I would like to use this bug as the placeholder/umbrella but for this contribution which will likely consist of multiple smaller contributions over the next months.

**Comment by scottslewis** [comment]
Hi Thomas.  

Congratulations on releasing 0.7.0...and Happy New Year!   It's nice to see ServiceFactory supported in iPopo, as it will help with implementing RSA.   

It's my desire to work on the ipopo RSA implementation more at the beginning of 2018, now that 0.7.0 is out.   I've done some initial work (i.e. creating org.osgi.services.remoteserviceadmin.* classes in python that define the RSA API...e.g. EndpointDescription, RemoteServiceAdminListener/Event, ImportReference/Registration, ExportReference/Registration, etc).

Before submitting these as pull requests however, I have a couple of things we should clarify:

1) RSA impls typically hook into the OSGi service registry with this OSGi-specified interface:

org.osgi.framework.hooks.service.EventListenerHook

Here's the javadocs:

https://www.knopflerfish.org/releases/5.2.0/docs/javadoc/org/osgi/framework/hooks/service/EventListenerHook.html

The first question to discuss:     This hook class doesn't yet exist in pelix.   Do we want to add it?    I could create the appropriate api and impl classes for pelix/ipopo, but as you can see by the signature it's actually quite close to service_listener API (i.e. ServiceEvent for first parameter).   Service events don't include the second parameter (BundleContext->ListenerInfo map) but for most RSA impls the second parameter isn't needed.   Another difference is that EventListenerHooks are setup via registration in the service registry rather than via BundleContext.add_service_listener.

For the initial RSA impl we could simply reuse ServiceListeners, or I could contribute an impl of EventListenerHook as an addition to pelix.   Thoughts about which would be preferred?

Second question:   what would you prefer for the package naming of the RSA API and impl?   Currently I have a new package called 'remoteserviceadmin' that is a peer to ipopo, remote, services,shell, etc.   Would you prefer 'rsa'? or some other package name?   Or a subpackage of services or some other package?

Please LMK what you think about these two questions and I'll start making pull requests...probably with event service hook first (if desired), and other dependencies, then API classes, then impl classes as separate pull requests

**Comment by tcalmant** [comment]
Hi Scott,

> Congratulations on releasing 0.7.0...and Happy New Year!

Thanks ! And Happy New Year :)

>For the initial RSA impl we could simply reuse ServiceListeners or I could contribute an impl of EventListenerHook as an addition to pelix

The main advantages of EventListenerHook are, like you said, that it's based on the whiteboard pattern, but also that *hooks* would be notified before *normal* service listeners; so I think it could be an interesting feature in some use cases of iPOPO itself.

This feature could be implemented quite easily, because all the work would be done inside the service registry and its associated event dispatcher (both in the ``pelix.internals.registry`` module).

So if you're OK to contribute this feature to Pelix, it will be really appreciated 👍 

> Second question: what would you prefer for the package naming of the RSA API and impl? Currently I have a new package called 'remoteserviceadmin' that is a peer to ipopo, remote, services,shell, etc. Would you prefer 'rsa'? or some other package name? Or a subpackage of services or some other package?

``pelix.remoteserviceadmin`` is too long (even with auto-completion :p ), and I don't think it should enter existing subpackages, as they are mainly for local usage.

At first, I though about ``pelix.remote.admin``, but maybe it could be confusing...
I'm fine with either ``pelix.rsa`` or maybe ``pelix.rsadmin``.

> I'll start making pull requests...probably with event service hook first (if desired), and other dependencies, then API classes, then impl classes as separate pull requests

That's great !
Also, using multiple pull requests will really help / ease tracking the addition of RSA 👍 

Thanks for all and again, Happy New Year :)

**Comment by scottslewis** [comment]
Hi Thomas,

One hitch about the EventListenerHook...the second parameter in EventListenerHook.event is a Map: i.e.

void event(ServiceEvent, Map<BundleContext,List<ListenerInfo>>)

Currently, the pelix Dispatcher adds service listeners with Dispatcher.add_service_listener(listener,specification, ldap_filter)

...with no passing of BundleContext, so the bundle context of the service listener  is not available at the fire_service_listener time.

I think that I could modify the signature of the Dispatcher.add_service_listener to:

def add_service_listener(bundle_context, listener, specification, ldap_filter) and then modify the implementation to store the bundle_context along with the listener (probably on modified version of class _Listener).   It seems to me that the only direct call to Dispatcher.add_service_listener is in BundleContext.add_service_listener, and so it would be easy to pass in BundleContext instance to Dispatcher.add_service_listener without breaking anything.   I wanted to check with you before doing so, however.   Any objections to adding a bundlecontext instance as the first argument to Dispatcher.add_service_listener to support the EventListenerHook semantics?

**Comment by scottslewis** [comment]
Also Thomas:  Would you please add me as a contributor to the ipopo repository?  I can then create/push an rsa branch and issue pull requests off that branch.   In the long term I think that will be easier for me than doing pull requests from a fork.   I promise to not commit or push directly to master.

**Comment by tcalmant** [comment]
Hi Scott,

No problem with adding the bundle context reference to the event dispatcher.

I've added you as contributor to the project, it's true it will ease this
kind of contribution.

**Comment by scottslewis** [comment]
Hi Thomas.

Question for you:   In my local workspace I've added on the pelix.rsa package, and have a pelix.rsa.remoteserviceadmin.py file...with a bundle activator declared.   If I try to create a framework with  a pelix.rsa.remoteserviceadmin bundle, however, I get error in importlib.import_module('pelix.rsa.remoteserviceadmin'):

Error installing bundle pelix.rsa:  No module named 'pelix.rsa'

I don't understand why it cannot be found by import_module, as it is present on filesystem right alongside (e.g.) pelix.shell.core and etc...which are found with no problem.   Is there something that has to be done for the import_bundle to 'know' about pelix.rsa.remoteserviceadmin?  

I'm launching the framework from another project (happen to be using pydev for this), but like I said it works fine with bundle 'pelix.shell.core' and others...but not pelix.rsa.remoteserviceadmin.   

Thanksinadvance.

**Comment by scottslewis** [comment]
Hi Thomas.

Please nevermind.   I discovered the project reference was just wrong and that's why the pelix.rsa package wasn't found.

**Comment by tcalmant** [comment]
Hi Scott,

> I've added several new modules in (new) rsa package to the rsa branch. I'm going to be working on these classes and some others over the next few days/weeks to fully implement RSA on ipopo. The code in rsa package compiles, but is not yet ready for testing, but I wanted to at least get these onto the rsa branch on github.

That's great news !

> What are your thoughts on next release timing for ipopo?

I don't have an exact time-frame for the next release, as I didn't have as much time as I'd like/need to properly work on iPOPO these last weeks.

On one hand, I have to add some tests on existing features and to add some documentation. Also, the TLS support in the Remote Shell seems to work, maybe I'll add it as an early-stage feature.

On the other hand, it might be interesting to publish a bug-fix release early (like this weekend), as an error was fixed around the log service. It would also give us more time for the following one.

> Please nevermind. I discovered the project reference was just wrong and that's why the pelix.rsa package wasn't found.

No worries, it happens often to me ;)

**Comment by scottslewis** [comment]
Hi Thomas,

I have a question for you about ipopo.    I'm creating component factory instances dynamically, via _IPopoService.instantiate(factory_name, name, properties).    I would, however, like to get at the name and properties values from within the component instance (e.g. in @Validate method for component).   How best to get the service_ref for the factory component instance and it's associated properties?

**Comment by scottslewis** [comment]
Hi Thomas,

Ok...I found PostRegistration decorator.   That seems to provide the service_ref....bingo!

Thanks.

**Comment by scottslewis** [comment]
One question about PostRegistration callbacks.   Given the name, I assume that these are called after the service is actually is put in the service registry.   Does this mean that some other thread (other than the one that calls ipoposervice.instantiate(factory_name, name, props) ) could begin calling methods on the newly created and registered service instance before PostRegistration actually completes.  Is that right?   Even if it is right it won't actually be a problem for me, as I'm only concerned with having PostRegistration called/completed before the ipoposervice.instantiate method returns.

About @Activate....I would do as you say, but I want to pass the props map into instantiate call (dynamic), rather than using @Property decorator (static).   That's why the PostRegistration is needed.

**Comment by scottslewis** [comment]
One other question:   what happens if the @PostRegistration callback throws and exception?   Is the component unregistered and instantiate throws?  or something else?

I guess what I'm getting to is that I need to do some initialization of the factory-created/instantiate instance...that could fail/exception, where the props map contains the initialization info (dynamic).

Thanks.

**Comment by scottslewis** [comment]
Here's one thought:   The @Validate decorator currently passes in the bundle context to the method implementation...e.g

@Validate
def _my_validate(self, bundle_context):
   # do stuff to validate

Would it be possible to enhance this validate notification so that the following signature could be supported?

@Validate
def _my_validate(self, bundle_context, props):
   # do stuff to validate

where props is a Map (the instance properties)

If that's difficult or impossible then perhaps an alternative would be

@Validate(component_context=True)
def _my_validate(self, component_context):
   # do stuff to validate

where the ComponentContext is passed in as the first arg rather than the bundle context...which can remain the default for the @Validate decorator...i.e. @Validate

With the component context, then the _my_validate method can/could also get at the instance properties (in the ComponentContext)

One reason I'm thinking this way is because of the @Activate annotation for DS, which optionally allows the bundle context, component context, and/or the Map of the instance properties (although the representation of these types is slightly different).

wdyt?

**Comment by tcalmant** [comment]
Hi Scott,

This seems to be a good idea, especially as it would mimic the behavior of DS.
Maybe this would require to open a new issue to have a specific discussion about this feature.

I think it would be better to use arguments on ``@Validate`` as it allows to be explicit on the arguments the method expects.
Note that it is complicated to have the ``@Validate`` decorator working both with and without arguments (without parenthesis). There is an [article](http://coderazzi.net/tnotes/python/decoratorsWithoutArguments.html) talking about the problem (last part).

Maybe it could be more interesting to have a new decorator, *e.g.* ``@ValidateExt()``, based on the same underlying configuration, but accepting arguments.
This would keep compatibility with the current code base and would be cleaner to implement and to test.

**Comment by scottslewis** [comment]
I've made a number of additions and have done some testing.  Things are not completed yet, and code/api documentation is needed along with test code, but I wanted to get working code in front of you or others...particularly for naming of modules and package, the overall structure of the classes etc.   

The latest of everything is on the rsa branch.

There is currently only one functioning distribution provider:  xmlrpc.  I'm working on a second one now.  I have not yet added the endpoint description discovery provider API.   That's soon and won't be nearly as complicated as the distribution provider API.

So, to use RSA with the xmlrpc provider the following bundles (at least) have to be added at startup:

    # Set the initial bundles
    bundles = ['pelix.ipopo.core', 'pelix.shell.core', 'pelix.shell.ipopo',
               'pelix.shell.console', 
               'pelix.rsa.remoteserviceadmin', 
               'pelix.http.basic', 
               'pelix.rsa.providers.dist.xmlrpc' ]

The pelix.rsa.remoteserviceadmin is the rsa impl itself, pelix.rsa.providers.dist.xmlrpc is the xmlrpc distribution provider, and pelix.http.basic is required by the xmlrpc provider (httpservice).

One question:  I called the package 'dist' for brevity but would you rather:  'distribution'?   

Once the above are started/active, there will be two new shell commands:

exportservice <service.id>
and
importservice

Before calling exportservice, however, you should install and start a test 'sample.hello' service with the necessary service properties.  For example:
```python
# iPOPO decorators
from pelix.ipopo.decorators import ComponentFactory, Provides, Instantiate
# Manipulates the class and sets its (unique) factory name
@ComponentFactory("hello-provider-factory")
# Indicate that the components will provide a service
@Provides("sample.hello")
# Tell iPOPO to instantiate a component instance as soon as the file is loaded
@Instantiate("hello-provider-auto", { 'service.exported.interfaces':'sample.hello', 
                                     'service.exported.configs':'ecf.xmlrpc.server', 
                                     'service.intents': ['osgi.async','osgi.basic'], 
                                     'ecf.xmlrpc.server.host':'localhost', 
                                     'osgi.basic.timeout':30000 })
# A component class must always inherit from object (new-style class)
class RemoteHelloImpl(object):

    def hello(self, name="world"):
        print("Service received:  Hello from {0} to Service".format(name))
        return "Hello from Service to "+name
```

Most of the service properties are there mostly for my testing (e.g. service intents, timeout, etc)

If you install this class and start/create a sample.hello service, and then lookup the helloimpl service.id, you can then give this command:

exportservice <helloimpl service.id>

and the rsa impl will export the service and print out the endpoint description created by the export.  For example:

```xml
$ exportservice 20
13:27:16.452;EXPORT_REGISTRATION;exp_svc_ref=ServiceReference(ID=20, Bundle=15, Specs=['sample.hello']);local_id=http://localhost:8181/xml-rpc;rs_id=http://localhost:8181/xml-rpc:1
---Endpoint Description---
<?xml version='1.0' encoding='cp1252'?>
<endpoint-descriptions xmlns="http://www.osgi.org/xmlns/rsa/v1.0.0">
	<endpoint-description>
		<property name="objectClass" value-type="String">
			<array>
				<value>sample.hello</value>
			</array>
		</property>
		<property name="remote.configs.supported" value-type="String">
			<array>
				<value>ecf.xmlrpc.server</value>
			</array>
		</property>
		<property name="service.imported.configs" value-type="String">
			<array>
				<value>ecf.xmlrpc.server</value>
			</array>
		</property>
		<property name="remote.intents.supported" value-type="String">
			<array>
				<value>osgi.basic</value>
				<value>osgi.async</value>
			</array>
		</property>
		<property name="endpoint.service.id" value="20" value-type="Long" />
		<property name="service.id" value="20" value-type="Long" />
		<property name="endpoint.framework.uuid" value="0fca5a4f-99df-4a44-b2d5-869d1967c775" value-type="String" />
		<property name="endpoint.id" value="6ef19ff9-e05f-41e7-af15-71d50d713336" value-type="String" />
		<property name="service.imported" value="true" value-type="String" />
		<property name="ecf.endpoint.id" value="http://localhost:8181/xml-rpc" value-type="String" />
		<property name="ecf.endpoint.id.ns" value="ecf.namespace.xmlrpc" value-type="String" />
		<property name="ecf.rsvc.id" value="1" value-type="Long" />
		<property name="ecf.endpoint.ts" value="1526995636452" value-type="Long" />
		<property name="ecf.exported.async.interfaces" value="*" value-type="String" />
		<property name="ecf.xmlrpc.server.host" value="localhost" value-type="String" />
		<property name="osgi.basic.timeout" value="30000" value-type="Long" />
	</endpoint-description>
</endpoint-descriptions>
---End Endpoint Description---
```
Here's the consumer of this service (setup with ldap filter to only have remotes injected with 'service.imported=*')
```python
# iPOPO decorators
from pelix.ipopo.decorators import ComponentFactory,Instantiate,Requires,Validate,Property

@ComponentFactory("remote-hello-consumer-factory")
@Requires("_helloservice", "sample.hello", False, False, "(service.imported=*)", False)
@Property("_name","name","Scott")
@Instantiate("hello-consumer")
class RemoteHelloConsumer(object):
    
    def __init__(self):
        self._helloservice = None
        self._name = None
        
    @Validate
    def _validate(self,ctxt):
        # call it!
        print("Consumer received: "+str(self._helloservice.hello(self._name)))
```
If you install the helloconsumer before you call importservice, then the required remote sample.hello will not exist yet (only the service exported), but if you issue this command:

$ importservice

then the helloconsumer should be validated and call the helloservice.hello method, resulting in an xmlrpc localhost request to the remoteservice impl and response.

This is how RSA works in 'manual' mode...i.e. with no topology manager and RSA service interfaces:  export_service, and import_service being invoked via console commands.   The exportservice and importservice commands replace the topology manager and discovery and it's possible to create custom topology manager,  discovery as well as new/other distribution providers (either from us or custom/private ones).   

There is a BasicTopologyManager (in pelix.rsa.topologymanagers.basic module) that for export does what you would expect for 'promiscuous' remote services...i.e. it exports every service with the required service properties (service.exported.interfaces).   Right now there's no network discovery...so it can't do the automatic import (yet).

Here's an architecture diagram from the rsa spec for your reference:

https://wiki.eclipse.org/EIG:Remote_Services_Admin#Architecture

Anyway...good progress.  Let me know if you have any thoughts about the organization and package/module/class naming in particular.

**Comment by scottslewis** [comment]
Lots more progress to report.

Refactored the package structure and the location of classes.  It's now organized like this:

(in pelix.rsa package):

providers/
    discovery/
        _init: EndpointEvent,EndpointEventListener...API,EndpointAdvertiser,EndpointSubscriber...sclasses
        etcd: Etcd-based impl of discovery API
    distribution/
        _init: DistributionProvider API,support classes
        xmlrpc: Implementation of DistributionProvider API based upon httpservice+xmlrpc
        I have another working distribution and discovery provider based upon py4j/osgiservicebridge.  It 
       depends upon python 3, however, so I think I'll distribute as separate from pelix rsa
topologymanagers/
    _init: TopologyManager superclass
        basic:  BasicTopologyManager(TopologyManager) class.  Implements 'promiscuous' topology manager (imports everything discovered, advertises for discovery everything exported, exports everything registered as remote service)

_init: RSA api, constants, support functions
edef:  EDEFReader and EDEFWriter from your remote code
endpointdescription:  EndpointDescription class (much by you) with support functions
remoteserviceadmin:  Impl of RemoteServiceAdmin class and RSA API (ExportReference/ExportRegistration,ImportReference/ImportRegistration, etc)
shell: RSA commands for the shell:  importservice,exportservice,listexports,listimports, etc

I've been testing/running with etcd server discovery for both advertise and subscribe and with xmlrpc and py4j distribution providers, and basictopologymanager and shell commands (no topology manager needed).   I'll begin writing integration tests similar to yours in tests.remote with appropriate changes as soon as I can.

I've also been adding documentation to both the API classes and the impl classes.   Not complete yet, but it's getting closer.   

To summarize:  this RSA impl has the basic structure for RSA as described here:

https://osgi.org/specification/osgi.cmpn/7.0.0/service.remoteserviceadmin.html

**Comment by tcalmant** [comment]
That's great !
I'll try it this weekend

**Comment by scottslewis** [comment]
Hi Thomas,

I've added quite a lot of documentation and a sample app.  Here's a way to give it a go:

1.  In samples directory, run top-level program:   run_rsa_xmlrpc.py.  Take a look at what bundles are there.
2.  To import sample hello server impl at console:
$ start samples.rsa.helloimpl
Note that the helloconsumer is already present (loaded by run), but it filters for remote services only '(service.imported=*)' so the helloimpl is not injected until explicitly imported below
3.  Do a service list:  $ sl   and note the service.id of the helloimpl instance.
4.  Type:  $ exportservice <service.id>
This should output to console...this is the RSA debugging output.  The output is
the RSA-created EndpointDescription
5.  Type$ listexports 
It should show you a table with one entry...i.e. the exportservice you just exported
If you type:  $ listexports <endpoint.id> it will show details about that exported service
6.  To import (and have the proxy created and injected into the helloconsumer) type:
$ importservice
Examine the output to the console.  It will show the import registration event into, as well
as the output from the remote call from helloconsumer to helloimpl (i.e. via xmlrpc distribution
provider).    Of course if you wish you can debug the helloconsumer to see when/how it's calling
via the helloimpl service proxy.  You can also examine the (now two) samples.hello service 
instances.

Feel free to play with the console commands, and of course with any of the code.   Also note that there is a topology manager to do all of the above automatically...it's in pelix.rsa.topologymanagers.basic.  I just wanted to expose you to the new things about rsa...e.g. the separation between service registration and export/import, the ability to replace the distribution or the discovery providers (or have multiple of either), the ability to see what's going on wrt import/export...either via console and/or programmatically.

I've got a few more things to add...e.g. the other distribution provider that I've created (based upon py4j), and I've got to add the support for service properties update, and I need to test with 2.7.   But it's pretty close now.

**Comment by scottslewis** [comment]
Hi Thomas.

I've been enhancing things further and have a small change to the instructions for running the export and import of a remote service given in previous comment.

here are the new instructions

In samples directory, run top-level program: run_rsa_xmlrpc.py. Take a look at what bundles are there.
To import sample hello server impl at console:
**$ start samples.rsa.helloimpl_xmlrpc**
Note that the helloconsumer is already present (loaded by run), but it filters for remote services only '(service.imported=*)' so the helloimpl is not injected until explicitly imported below
Do a service list: 
**$ sl** 
and note the service.id of the helloimpl instance.
Type: 
**$ exportservice <service.id>**
This should output to console...this is the RSA debugging output. The output is
the RSA-created EndpointDescription
Type:
**$ listexports**
It should show you a table with one entry...i.e. the exportservice you just exported
If you type: $ listexports <endpoint.id> it will show details about that exported service
To import (and have the proxy created and injected into the helloconsumer) type:
**$ importservice**
Examine the output to the console. It will show the import registration event into, as well
as the output from the remote call from helloconsumer to helloimpl (i.e. via xmlrpc
provider). Of course if you wish you can debug the helloconsumer to see when/how it's calling
via the helloimpl service proxy. You can also examine the (now two) samples.hello service
instances

I've also decided to include the py4j distribution (and discovery) provider if desired.   Because of dependency on concurrent api, it requires Python 3.x.   It also depends upon the osgiservicebridge code (both Python and Java).   See the [Py4jRemoteServiceProvider](https://github.com/ECF/Py4j-RemoteServicesProvider) project repo.  I've been testing this and it's pretty cool...it can use Google protobuf for optimized serialization/deserialization of args and return values, and since ECF RS now supports [OSGi Async Remote Services](https://osgi.org/specification/osgi.cmpn/7.0.0/service.remoteservices.html#d0e1407), service developers can use this to make async calls between Java and Python simply by declaring Java service interfaces that use CompletableFuture, Future, or Promise.  Everything else is done by the distribution provider!   You can look at the samples _py4j.py files to get an idea how it works from the Python side.

I would like to begin the process of merging into master.  I'm not sure what your release timeline is, but since ECF is having it's release at end of June, it might be nice...for promotion...to have a release of this addition sometime around that time?

**Comment by tcalmant** [comment]
Hi Scott,

That's great :)
I tried RSA last weekend and it was working as intended.
I'll be looking a bit further into it in order to see how to provide a an exporter based on JSON-RPC, to have a better view on how it works.

The Py4J provider could be tried with the [unofficial backport of ``futures``](https://pypi.org/project/futures/) on Python 2.7.
I'll try to do the tests this weekend, if I find the time.

I'm OK with a release at the same time as ECF.
It will give me some time to add a bit of documentation on the website.

**Comment by scottslewis** [comment]
Hi Thomas.

FWIW: I forked the Java Cohorte code some time ago...to keep up with enhancements in the RSA java impl:  https://github.com/scottslewis/cohorte-remote-services.    I can/could enhance them even further (the java distribution provider API has been enhanced recently to support OSGi Async Remote Services spec).   I would be happy to update this provider and contribute back any older changes.   If this was done (in combination with a python distribution provider) it would be yet another that would immediately support Python<->Java via remote services/RSA.
